adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala

```diff
@@ -327,35 +338,44 @@ private[rdd] class BEDParser extends FeatureParser {
  * Specification:
  * http://www.genome.ucsc.edu/FAQ/FAQformat.html#format12
  */
-private[rdd] class NarrowPeakParser extends Serializable with Logging {
+private[rdd] class NarrowPeakParser extends FeatureParser {
 
-  def parse(line: String,
-            stringency: ValidationStringency): Seq[Feature] = {
-    val fields = line.split("\t")
-    def hasColumn(n: Int): Boolean = {
-      fields.length > n && fields(n) != "."
+  def isHeader(line: String): Boolean = {
+    line.startsWith("#") || line.startsWith("browser") || line.startsWith("track") || line.isEmpty
+  }
+
+  override def parse(line: String,
+                     stringency: ValidationStringency): Option[Feature] = {
+
+    // skip header lines
+    if (isHeader(line)) {
+      return None
     }
 
-    // skip empty or invalid lines
+    val fields = line.split("\t")
+
+    // check for invalid lines
     if (fields.length < 3) {
       if (stringency == ValidationStringency.STRICT) {
-        throw new IllegalArgumentException("Empty or invalid NarrowPeak line: %s".format(line))
+        throw new IllegalArgumentException("Invalid NarrowPeak line: %s".format(line))
       } else if (stringency == ValidationStringency.LENIENT) {
-        log.warn("Empty or invalid NarrowPeak line: {}", line)
+        log.warn("Invalid NarrowPeak line: {}", line)
       }
-      return Seq()
+      return None
+    }
+
+    def hasColumn(n: Int): Boolean = {
+      fields.length > n && fields(n) != "."
     }
-    val fb = Feature.newBuilder()
-    fb.setContigName(fields(0))
 
-    // Peak files are 0-based space-coordinates, so conversion to
-    // our coordinate space should mean that the values are unchanged.
-    fb.setStart(fields(1).toLong)
-    fb.setEnd(fields(2).toLong)
+    val f = Feature.newBuilder()
+      .setContigName(fields(0))
+      .setStart(fields(1).toLong) // NarrowPeak ranges are 0-based
```

Comment:
We should try/catch around the `fields(x).toLong`/`fields(x).toDouble` calls, both here and below and above and everywhere. We could factor it out like so:

```
def checkAndSet[T](field: String,
  convFn: String => T,
  setFn: T => Unit,
  stringency: ValidationStringency) {
  try {
    val t = convFn(field)
    setFn(t)
  } catch {
    case e: Throwable => {
      if (stringency == ValidationStringency.LENIENT) {
        log.warn("Failed to convert %s.".format(field))
      } else if (stringency == ValidationStringency.STRICT) {
        throw new IllegalArgumentException("Setting field from %s failed with %s.".format(
          field,
          e.msg))
      }
    }
  }
}
```

Perhaps we factor this out into a trait:

```
sealed trait FeatureConverterCanBeValidated {
  val stringency: ValidationStringency
  
  def validateOrNone(...): Option[Feature]

  def checkAndSet[T](field: String,
    convFn: String => T,
    setFn: T => Unit)
}
```

All of the classes in here could be turned into case classes that extend that trait. What do you think? Should you proceed, the trait name I proposed is terrible, so please change it.